### DISCUSSION TITLE
How to cache/shorten cross-compilation of windows builds?

### DISCUSSION THREAD

**Discussion by mpizenberg** [PyO3/maturin#1953] [Q&A] [answered]
I'm currently cross-compiling for Windows with `maturin build --release --target x86_64-pc-windows-msvc --zig` and adding the `generate-import-lib` feature to pyo3. This combination makes my GitLab builds timings increase by roughly 10 minutes. What is the main culprit for that? And is it possible to cache whatever happens during these 10 minutes? For now I'm caching `.cargo/` and `target/` but it does not seem to make a difference.

This question might be related to https://github.com/PyO3/maturin/discussions/1947

**Comment by messense** [comment]
You can try to cache `~/.cache/cargo-xwin`.

**Reply by mpizenberg** [reply]
I can only cache paths within the repo in gitlab. For cargo I'm assigning `CARGO_HOME` to `$CI_PROJECT_DIR/.cargo`, is there a similar env variable for cargo-xwin?

**Reply by mpizenberg** [reply]
found it :) https://github.com/rust-cross/cargo-xwin?tab=readme-ov-file#customization
Thanks will try

**Reply by messense** [reply]
`XWIN_CACHE_DIR` probably won't work in `maturin` because it was a clap `env` option in `cargo-xwin`, the way we call cargo-xwin in `maturin` is a bit different which does not go through `clap`.

So I think we need to add support for it in https://github.com/PyO3/maturin/blob/b209c4b7b10f5bcf6141c85ed26999401568b2fa/src/compile.rs#L282-L288, pull requests are welcome.

**Reply by mpizenberg** [reply]
Ok, thanks for the pointer. I doubt I can tackle this in short term. Hopefully somebody will.

**Reply by NfNitLoop** [reply]
Thanks for the info! I'm doing a build inside of Docker and this will allow me to use a cache mount to speed things up dramatically on my slow internet. 😊 

> I can only cache paths within the repo in gitlab

@mpizenberg Maybe as a workaround your GitLab build can just add a symlink from `~/.cache/cargo-xwin` into a directory in your repo?

**Reply by messense** [reply, marked as answer]
Maybe you can try set `XDG_CACHE_HOME` env var to a path in the repo since `cargo-xwin` uses `dirs` crate which follows the XDG base directory spec.

**Reply by mpizenberg** [reply]
Thanks @NfNitLoop and @messense for the ideas. I haven't tested the symlink idea, but I can confirm that setting `XDG_CACHE_HOME` to `$CI_PROJECT_DIR/.cache` and caching `.cache/` works in GitLab :)

**Reply by messense** [reply]
I'm adding support for `cargo-xwin` env vars in #1961 to make things easier.
